pkg/proxy/iptables/proxier.go

```diff
@@ -1126,24 +1126,22 @@ func (proxier *Proxier) syncProxyRules() {
 				)
 
 				destChain := svcXlbChain
-				// We have to SNAT packets to external IPs if externalTrafficPolicy is cluster.
+				// We have to SNAT packets to external IPs if externalTrafficPolicy is cluster
+				// and the traffic is NOT Local. Local traffic coming from Pods and Nodes will
+				// be always sent directly to the corresponding Service, so no need to SNAT
 				if !svcInfo.OnlyNodeLocalEndpoints() {
 					destChain = svcChain
-					writeLine(proxier.natRules, append(args, "-j", string(KubeMarkMasqChain))...)
+					if proxier.localDetector.IsImplemented() {
+						// This masquerades off-cluster traffic to a External IP. The idea
+						// is that you can establish a static route for your Service range,
+						// routing to any node, and that node will bridge into the Service
+						// for you.  Since that might bounce off-node, we masquerade here.
+						// If/when we support "Local" policy for VIPs, we should update this.
+						writeLine(proxier.natRules, proxier.localDetector.JumpIfNotLocal(args, string(KubeMarkMasqChain))...)
+					}
 				}
+				writeLine(proxier.natRules, append(args, "-j", string(destChain))...)
 
-				// Allow traffic for external IPs that does not come from a bridge (i.e. not from a container)
-				// nor from a local process to be forwarded to the service.
-				// This rule roughly translates to "all traffic from off-machine".
-				// This is imperfect in the face of network plugins that might not use a bridge, but we can revisit that later.
-				externalTrafficOnlyArgs := append(args,
-					"-m", "physdev", "!", "--physdev-is-in",
-					"-m", "addrtype", "!", "--src-type", "LOCAL")
```

Comment:
Switching to the local traffic detector is probably not currently a compatible replacement for this rule, since we don't currently have the physdev-based local traffic detector, right?

Reply:
I think that the discussion is if this rule is correct, that it seems it is not, because CNIs not using bridges are not affected ...
what we want to do is to SNAT all the non-local traffic that is what the local detector should do, independently of the implementation

Reply:
I am not sure this is quite correct.

Before this change, we try to:
1) always MASQ
2) send all from-off-node traffic to the service chain
3) send from-on-node traffic to the service chain IFF the dst is "local"
4) send any other from-on-node traffic to the network

(4) matters because the IP presumably might do some processing/telemtry/whatever and we don't want to short-circuit that
(3) matters because the result of (4) will be wrong if the IP is local

Now, most CNIs, including kubenet (I think) don't use the bridge any more so we effectively do:
1) always MASQ
2) send all traffic to the service chain

What you do after this change is:
1) MASQ from-off-node traffic if policy is not local-only (seems ok)
2) send all traffic to the service chain

Your (1) seems like an improvement.
Your (2) seems like not worse than what has ACTUALLY been happening.

This field is rather under-specified and, in thinking about it, it was probably wrong before.  So OK.  I think I am in.  I have a nit above, but I am happy to see physdev go away